include/cuco/detail/priority_queue.inl

```diff
@@ -0,0 +1,197 @@
+priority_queue<T, Compare, Allocator>::~priority_queue()
+{
+  std::allocator_traits<int_allocator_type>::deallocate(int_allocator_, d_size_, 1);
+  std::allocator_traits<size_t_allocator_type>::deallocate(size_t_allocator_, d_p_buffer_size_, 1);
+  std::allocator_traits<t_allocator_type>::deallocate(
+    t_allocator_, d_heap_, node_capacity_ * node_size_ + node_size_);
+  std::allocator_traits<int_allocator_type>::deallocate(
+    int_allocator_, d_locks_, node_capacity_ + 1);
+}
+
+template <typename T, typename Compare, typename Allocator>
+template <typename InputIt>
+void priority_queue<T, Compare, Allocator>::push(InputIt first, InputIt last, cudaStream_t stream)
+{
+  constexpr int block_size = 256;
+
+  const int num_nodes  = static_cast<int>((last - first) / node_size_) + 1;
+  const int num_blocks = std::min(64000, num_nodes);
+
+  detail::push_kernel<<<num_blocks, block_size, get_shmem_size(block_size), stream>>>(
+    first,
+    last - first,
+    d_heap_,
+    d_size_,
+    node_size_,
+    d_locks_,
+    d_p_buffer_size_,
+    lowest_level_start_,
+    compare_);
+
+  CUCO_CUDA_TRY(cudaGetLastError());
+}
+
+template <typename T, typename Compare, typename Allocator>
+template <typename OutputIt>
+void priority_queue<T, Compare, Allocator>::pop(OutputIt first, OutputIt last, cudaStream_t stream)
+{
+  constexpr int block_size = 256;
+  const int pop_size       = last - first;
+
+  const int num_nodes  = static_cast<int>(pop_size / node_size_) + 1;
+  const int num_blocks = std::min(64000, num_nodes);
+
+  detail::pop_kernel<<<num_blocks, block_size, get_shmem_size(block_size), stream>>>(
+    first,
+    pop_size,
+    d_heap_,
+    d_size_,
+    node_size_,
+    d_locks_,
+    d_p_buffer_size_,
+    lowest_level_start_,
+    node_capacity_,
+    compare_);
+
+  CUCO_CUDA_TRY(cudaGetLastError());
+}
+
+template <typename T, typename Compare, typename Allocator>
+template <typename CG, typename InputIt>
+__device__ void priority_queue<T, Compare, Allocator>::device_mutable_view::push(CG const& g,
+                                                                                 InputIt first,
+                                                                                 InputIt last,
+                                                                                 void* temp_storage)
+{
+  const detail::shared_memory_layout<T> shmem =
+    detail::get_shared_memory_layout<T>((int*)temp_storage, g.size(), node_size_);
+
+  const auto push_size = last - first;
+  for (std::size_t i = 0; i < push_size / node_size_; i++) {
+    detail::push_single_node(g,
+                             first + i * node_size_,
+                             d_heap_,
+                             d_size_,
+                             node_size_,
+                             d_locks_,
+                             lowest_level_start_,
+                             shmem,
+                             compare_);
```

Comment:
`push_single_node`, `push_partial_node`, and related utilities should be member functions of `device_mutable_view`.

Reply:
The same as `pop_single_node` and `pop_partial_node`